client/components/swipeable/index.js

```diff
@@ -0,0 +1,212 @@
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
+			}
+
+			savedUpdateLayout.current();
+		}, [ enabled, currentPageIndex ] );
+
+		// fire when the window resizes
+		useEffect( () => {
+			if ( ! enabled ) {
+				return;
+			}
+
+			const onResize = () => savedUpdateLayout.current( false );
+			window.addEventListener( 'resize', onResize );
+			window.addEventListener( 'orientationchange', onResize );
+			return () => {
+				window.removeEventListener( 'resize', onResize );
+				window.removeEventListener( 'orientationchange', onResize );
+			};
+		}, [ enabled ] );
+	}
+
+	const updateEnabled = hasDynamicHeight && numPages > 1;
+
+	useUpdateLayout( updateEnabled, currentPage, ( hasTransitionDuration = true ) => {
+		const offset = getOffset( currentPage );
+		const transform = {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: hasTransitionDuration ? `300ms` : `0ms`,
+		};
+
+		let height = { height: null };
+		const targetHeight = pagesRef.current?.querySelector( '.is-current' )?.offsetHeight;
+
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+
+		setPagesStyle( { ...pagesStyle, ...height, ...transform } );
+	} );
+
+	useEffect( () => {
+		if ( currentPage >= numPages ) {
+			onPageSelect( numPages - 1 );
+		}
+	}, [ numPages, currentPage, onPageSelect ] );
+
+	const getDragPosition = ( event ) => {
+		if ( event.hasOwnProperty( 'clientX' ) ) {
+			return { x: event.clientX, y: event.clientY };
+		}
+
+		if ( event.targetTouches[ 0 ] ) {
+			return { x: event.targetTouches[ 0 ].clientX, y: event.targetTouches[ 0 ].clientY };
+		}
+
+		const touch = event.changedTouches[ 0 ];
+		return { x: touch.clientX, y: touch.clientY };
+	};
+
+	const handleDragStart = ( event ) => {
+		const position = getDragPosition( event );
+		setDragStartData( position );
+		setPagesStyle( { ...pagesStyle, transitionDuration: `0ms` } ); // Set transition Duration to 0 for smooth dragging.
+		setIsDragging( true );
+	};
+
+	const handleDragEnd = ( event ) => {
+		if ( ! isDragging ) {
+			return; // End early if we are not dragging any more.
+		}
+
+		const THRESHOLD_OFFSET = 100; // Number of pixels to travel before we trigger the slider to move to the desired slide.
+		const dragPosition = getDragPosition( event );
+		const delta = dragPosition.x - dragStartData.x;
+
+		const hasMetThreshold = Math.abs( delta ) > THRESHOLD_OFFSET;
+
+		let newIndex = currentPage;
+		if ( delta < 0 && hasMetThreshold && numPages !== currentPage + 1 ) {
+			newIndex = currentPage + 1;
+		}
+
+		if ( delta > 0 && hasMetThreshold && currentPage !== 0 ) {
+			newIndex = currentPage - 1;
+		}
+		const offset = getOffset( newIndex );
+		setPagesStyle( {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: `300ms`,
+		} );
+		onPageSelect( newIndex );
+		setDragStartData( {} );
+		setIsDragging( false );
+	};
+
+	const handleDrag = ( event ) => {
```

Comment:
This may not be as important as other perf-related comments, but we should look into memoizing (via the `useCallback` hook) these functions, which are otherwise added/removed as event handlers for every render:

- `getDragPosition`
- `handleDragStart`
- `handleDrag`
- `handleDragEnd`

Reply:
From my limited performance testing, all the animations seem to run well. (58 fps) I am not sure we need to add the useCallback hooks.

Reply:
For sure this could be a bit of premature optimisation, but since you're mentioning FPS, what device did you test it on?

I'm asking because for this kind of testing, we should make sure that we can hit (ideally) 60fps on a low-end device — testing on the machines that we use to code daily is not a good reference for the average use (cc @sgomes )

Reply:
It's always a good idea to measure performance and make sure that you're making an informed decision, and as @ciampo points out, doing so on lower-end devices that we want to target as our baseline (or, alternatively, using something like the CPU slowdown feature in Chrome DevTools). This avoids needlessly complicating the code, but making sure that out bases are well covered.

That said, adding and removing event handlers on every render is probably not a great practice, even if React has its own event handling layer that sits atop the DOM. I'm a bit rusty on my React hooks, but if this is achievable with minimal code and memory overhead, it's probably a good idea to do a little less work on every render.

Do let me know if you'd like me to dig deeper and do some runtime performance testing on this!